httpclient/http_client.go

```diff
@@ -95,18 +107,46 @@ func (hc *HTTPClient) DoRequest(ctx context.Context, pld HTTPRequest) (*http.Res
 		}
 	}
 
-	resp, err := hc.Client.Do(req)
+	return hc.Client.Do(req)
+}
+
+func hasSuccessCode(resp *http.Response) bool {
+	return resp.StatusCode >= 200 && resp.StatusCode < 300
+}
+
+type RequestFuncion func(ctx context.Context, pld HTTPRequest) (*http.Response, error)
+type ParserFunction[T any] func(resp *http.Response) (T, error)
+type ErrorParserFunction func(statusCode int, body []byte) error
+
+func RequestModel[T any](
+	ctx context.Context,
+	req RequestFuncion,
+	pld HTTPRequest,
+	parser ParserFunction[T],
+	errorParser ErrorParserFunction,
+) (model T, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("requestModel panic: %v", r)
+		}
+	}()
+	resp, err := req(ctx, pld)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
 	if err != nil {
-		return nil, err
+		return
 	}
 
-	if hasSuccessCode(resp) {
-		return resp, nil
+	if !hasSuccessCode(resp) {
+		err = handleErrorResponse(resp, errorParser)
+		return
 	}
 
-	return nil, HttpClientError{resp}
-}
+	model, err = parser(resp)
+	if err != nil {
+		return
+	}
 
-func hasSuccessCode(resp *http.Response) bool {
-	return resp.StatusCode >= 200 && resp.StatusCode < 300
+	return
```

Comment:
It's a minor, but I'm not in favour of naked returns. 

I'm not gonna force it tho, just sayin. Unless someone agrees with me? 😄

Reply:
Generally, I try to avoid naked returns. However, there is one exception when 'standard' returns can't be applied: it's the case (this case) when I want to return an error based on deferred recovery.

There might be another question: why did I decide to write the 'defer recover' part? 
There are two factors:

- This function makes an HTTP call to a remote server, and during that, unpredictable events might occur (such as timeouts or socket closures), so it's better to 'try-catch' potential errors.
- This function utilizes other functions (`parser` and `errorParser`) passed as arguments, so I can't assume that they will work without failures."